cmd/libsnap-confine-private/device-cgroup-support.c

```diff
@@ -107,6 +111,322 @@ static void _sc_cgroup_v1_attach_pid(sc_device_cgroup *self, pid_t pid) {
     sc_dprintf(self->v1.fds.cgroup_procs_fd, "%i\n", getpid());
 }
 
+/**
+ * sc_cgroup_v2_device_key is the key in the map holding allowed devices
+ */
+struct sc_cgroup_v2_device_key {
+    uint8_t type;
+    uint32_t major;
+    uint32_t minor;
+} __attribute__((packed));
+typedef struct sc_cgroup_v2_device_key sc_cgroup_v2_device_key;
+
+/**
+ * sc_cgroup_v2_device_value holds the value stored in the map
+ *
+ * Note that this type is just a helper, the map cannot be used as a set with 0
+ * sized value so we always store something in it (specifically value 1) in the
+ * map.
+ */
+typedef uint8_t sc_cgroup_v2_device_value;
+
+static void _sc_cgroup_v2_attach_pid(sc_device_cgroup *self, pid_t pid) {
+    /* nothing to do here, the device controller is attached to the cgroup
+     * already, and we are part of it */
+}
+
+#ifdef ENABLE_BPF
+static int load_devcgroup_prog(int map_fd) {
+    // Basic rules about registers:
+    // r0    - return value of built in functions and exit code of the program
+    // r1-r5 - respective arguments to built in functions, clobbered by calls
+    // r6-r9 - general purpose, preserved by callees
+    // r10   - read only, stack pointer
+    // Stack is 512 bytes.
+    //
+    // The function declaration implementing the program looks like this:
+    // int program(struct bpf_cgroup_dev_ctx * ctx)
+    // where *ctx is passed in r1, while the result goes to r0
+    //
+    /* just a placeholder for map value */
+    uint8_t map_value __attribute__((unused));
+    // where the value is 1 byte, but effectively ignored at this time. We are
+    // using the map as a set, but 0 sized key cannot be used when creating a
+    // map.
+    size_t key_start = 17;
```

Comment:
I cannot mentally link the comment to the variable declaration. Could you please make it more explicit, also explaining where the value 17 comes from?

Reply:
The comments got mixed up during all the rebases I did. I'll try to make it more useful.

Reply:
Did this ever get done? I can't easily understand where 17 comes from - the comment below says '4 byte-aligned address -1' but 17 would appear (at least to my naive understanding) as a 4-byte-aligned address +1.

Reply:
This comes from the hack I described just below. The idea is that word-sized operations needs to access word aligned values on the stack. So the actual map key structure is positioned to start at sp - 17, such that it fits entirely (struct is 1+4+4 bytes). Since it is packed, there's no gaps in between the fields. The structure starts with `uint8_t type` (at sp-17 offset), but the next uint32_t sized fields which are accessed by instructions operating on word sized arguments start at `sp-16`, and are thus properly aligned.

Another solution to this (maybe a bit more comprehensive?) would be to move the type field to the end, and eg. start the structure at an aligned offset at the stack (eg. sp-16), thus major/minor are at aligned offsets, while he type field is byte sized and the code `BPF_ST_MEM(BPF_B,..)` can still work. 

For refernce, the BPF code generated by llvm did something even worse as it would shift and fill respective fields byte by byte.